### ISSUE TITLE
[RFC] Allowing Users to Simply Opt-in to Features

### ISSUE THREAD

**Issue by rossriley** [bolt/bolt#6530]
Obviously the main driver of this is to allow advanced users to move onto the new storage system so I've used a rough outline of how this could work in this RFC but the same principle will apply to any experimental feature that we want to allow users to opt in to.

## Step 1 - Enabling the Experimental Feature Service

The configuration for this is done in the `.bolt.yml` file. The syntax can look something like this.

```yml
application: MyApp\Application
paths:
    web: public
    themebase: public/theme
    files: public/files
    view: public/bolt-public/view
services:
  -  Bolt\Provider\FeatureProvider:
        feature.storage_twig_new: true
```

Listing the service here enables users to opt into the load of this provider. Whilst we can ship this in `src/Bolt/Provider/FeatureProvider.php` it will not be loaded by default in a standard application it needs the user to put the class into their Bolt config to get it loaded.

The services definition in `.bolt.yml` optionally takes an array of values too, so we can use these to toggle individual features by setting to true. So `feature.storage_twig_new: true` allows us to opt into this feature but there could also be a list of several defined at this point.

## Step 2 - Enabling the Experimental Feature

Because we now have a way for the user to opt into features we can optionally extend the default services in our new provider. For instance the FeatureProvider class may look a little like this...

```php
<?php

namespace Bolt\Provider;

use Silex\ServiceProviderInterface;
use Silex\Application;
use Twig_Environment;

/**
 *
 */
class FeatureProvider implements ServiceProviderInterface
{

    public function register(Application $app)
    {
        if ($app['feature.storage_twig_new']) {
            $app['twig'] = $app->share(
                $app->extend(
                    'twig',
                    function (Twig_Environment $twig, $app) {
                        $twig->addTokenParser(new Experimental\SetcontentTokenParser());
                    }
                )
            );
        }


    }

    public function boot(Application $app)
    {
    }
}
```

We wrap the extension definition in an if block to only modify it the feature has been toggled on.

## Possible Risks

When implementing feature toggles there is always an extra layer of complexity that is being introduced, feasibly one user will see different behaviour depending on which features are on. But if we try and centralise the available features and define the extensions in a single provider then at least we have a very easy-to-scan definition of what can be opted into.

When we want to adopt these features into core,  we can just remove the entire if block from the FeatureProvider and the user's config will just do nothing any more.

If we want to abandon a feature then we would need to either deprecate and trigger warnings in the if block, or alternatively for features that are more experimental but may cause breakages when removed we could perhaps throw an exception that warned the developer that the flag they had set eg: 

```yml
services:
  -  Bolt\Provider\FeatureProvider:
        feature.use_blade_instead_of_twig: true
```

Could be met with a 'This experimental feature is no longer available' error that explains that it needs to be removed.

**Comment by CarsonF** [comment]
Here's the problem with that: The flag is being checked (invoked) inline / at registration time. The parameters given to the service provider `feature.storage_twig_new` are set after the if statement has checked it's value. It needs to be done inside of the closure. Like this:
```php
$app['twig'] = $app->share($app->extend('twig', function ($twig, $app) {
    if ($app['feature.storage_twig_new']) {
        $twig->addTokenParser(new Experimental\SetcontentTokenParser());
    }
    return $twig;
}));
```

If you look at any of Silex's SPs none of them do logic based on `$app` values directly in the `register` method, they are all in closures of the services. The only exception there is appending to arrays.

----

I don't think having them centralized in one SP is a good idea either. As note above they have to be inside the extend closures so you can't do one if statement for every service you want to change. It has to be an if statement for every service. Each extend is verbose and does add a little more overhead. Besides this it just seems to make more sense to me to have them next to the feature they augmenting / replacing.

Additionally it may not be possible to change what we in the dependency container. For example, I need to know inside `Controller\Base::render` whether to return a `TemplateResponse` or `TemplateView`. Yes it could be abstracted to DI, but that's potentially more refactoring on code we are trying to replace.

---

I think a lot of these flags could go directly into the Config. I think there are only a few edge cases where we couldn't check config for these flags. Notably, anything related to Config as that would be chicken & egg problem.

`.bolt.yml` should just be for bootstrapping (not to be confused with _boot_ phase). There's no need for feature flags to add new services, because it's a new service it can be unconditionally registered. Since all feature checks happen lazily, it's ok for them to depend on the Config service.

---

I can see value for having both _compatibility_ and _experimental_ flags. The first allows users to update their code on their terms, and deprecated value will throw a warning now and change will be forced on them in the next version and the deprecation check removed. The latter should allow us to make changes without worry of BC. See Symfony's blog post on it: http://symfony.com/blog/experimental-features. If we decided to remove an experimental feature, I wouldn't want to have to keep a deprecation check for it as that's like BC. 

I don't think experimental should be an excuse for buggy code either.

I agree that this makes debugging harder for us, and adds complexity for users. I would caution us to only use these as a last resort. Every flag we create means extensions could have to handle **both** values it could be.

**Comment by bobdenotter** [comment]
I also think these should be in Config. 

Other than that, I like the plan. 👍

**Comment by GawainLynch** [comment]
> I think a lot of these flags could go directly into the Config

I _really_ don't. `Config` is on our hit-list, so there is only a matter of time until "Chicken vs. Egg — The Final Showdown" lands, and we're having fun conversations in #development about how to get out of "this" hole.

> I can see value for having both compatibility and experimental flags

Not sure I see where you're going with the `compatibility` flag, isn't that normal SoP?

As for "experimental" features … extensions should be very cautious about supporting their products, as there is and should be **zero BC guarantees** on experimental features.

So when experimental features disappear, that fact appears in the release notes, and all traces go from our code base  … "Your app crashed after upgrade? Well you had `experimental` turned on, and you didn't read the release notes" :wink:

**Comment by rossriley** [comment]
Yes, reason I suggested the services method was because I thought we said config was a no-go, at the stage where these things are set I think config objects are still all just proxy objects and I thought the plan was not to load those until we had a request?

I really don't envision us having more than a handful of these flags available at any time and essentially it's targeted at stuff that is new, almost certainly going to end up as 'the approved way' to do something but that we can't just switch on immediately because of the breaking nature of what we've done.

So what we need is a way for users to gradually transition to a new approach whilst making sure that everybody else's existing code continues to work unaffected. So the risks section was very much theoretical considerations but in reality I'm not convinced that we'll ever need to pull one of these unless we have some major PHP6 moment in the future.

**Comment by CarsonF** [comment]
> Not sure I see where you're going with the compatibility flag, isn't that normal SoP?

The idea was for places where we can't maintain BC. Going back to the `TemplateView` feature, we don't know if the user will try to call Response methods on the return value of `$this->render()`. But instead of having another class to migrate to, like `NewFrontend`, we would just have a config option for this.
It _IS_ different than `experimental`. It's not a "cutting edge" feature that could be changed/removed. This is a for sure thing and the old way will be removed in next major release.

I could see `experimental` flags turning into `compatibility` flags as the feature becomes stable and adoption is encouraged instead of cautioned.

**Comment by bobdenotter** [comment]
Resolved by #6443
